### PR TITLE
Integrate the Google Play Billing Library into engine

### PR DESCRIPTION
**Pull request by qiuguohua**
Re: #
https://developer.android.com/google/play/billing/integrate
https://developer.android.com/reference/com/android/billingclient/api/BillingClient?_gl=1*11tcdon*_up*MQ..*_ga*MjcyNzYxMzIuMTczMTM3NTA0MQ..*_ga_6HH9YJMN9M*MTczMTM3NTA0MS4xLjAuMTczMTM3NTA0MS4wLjAuNzA1NTQwMDMy
### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: cocos/core/platform/billing/billing.ts [hunk 1]
```diff
@@ -0,0 +1,804 @@
...
+    /**
+     * @en
+     * This field is deprecated.
+     * See SERVICE_UNAVAILABLE which will be used instead of this code.
+     *
+     * @zh
+     * 这个字段已经废弃。
+     * 看看SERVICE_UNAVAILABLE将使用哪一个来代替此代码。
+     */
+    SERVICE_TIMEOUT = -3,
```
Comment (qiuguohua) [inline cocos/core/platform/billing/billing.ts]: This field is marked as deprecated in the google docs, does it need to be removed?

File: cocos/core/platform/billing/billing.ts [hunk 2]
```diff
@@ -0,0 +1,804 @@
...
+    [BillingEventType.EXTERNAL_OFFER_AVAILABILITY_RESPONSE]: (result: native.BillingResult) => void
+    [BillingEventType.ALTERNATIVE_BILLING_ONLY_INFORMATION_DIALOG_RESPONSE]: (result: native.BillingResult) => void
+    [BillingEventType.EXTERNAL_OFFER_INFORMATION_DIALOG_RESPONSE]: (result: native.BillingResult) => void
+    [BillingEventType.IN_APP_MESSAGE_RESPONSE]: (result: native.InAppMessageResult) => void
+}
+
+export class GooglePlayBilling {
+    private _eventTarget: EventTarget = new EventTarget();
+    constructor () {
+        if (!JSB || !jsb.billing) {
```
Comment (qiuguohua) [inline cocos/core/platform/billing/billing.ts]: Return if non-JSB or jsb.billing is null, only in android platform jsb.billing is not null

File: cocos/native-binding/index.ts [hunk 3]
```diff
@@ -1476,4 +1476,349 @@ export declare namespace native {
...
+         * @zh 唯一标识给定商品和用户对的购买的令牌。
+         */
+        readonly purchaseToken: string;
+        /**
+         * @en String containing the signature of the purchase data that was signed with the private key of the developer.
+         * @zh 包含使用开发者私钥签名的购买数据签名的字符串。
+         */
+        readonly signature: string;
+
+        readonly toStr: string;
```
Comment (qiuguohua) [inline cocos/native-binding/index.ts]: The name should be toString, but ts has an internal toString, so it's changed to toStr.

File: native/cocos/platform/android/modules/google_play/billing/JniBillingHelper.cpp [hunk 4]
```diff
@@ -0,0 +1,583 @@
...
+void JniBillingHelper::launchBillingFlow(const std::vector<ProductDetails*>& productDetailsList, const std::string& selectedOfferToken) {
+    if (productDetailsList.empty()) {
+        return;
+    }
+    auto* env = JniHelper::getEnv();
+    const int size = productDetailsList.size();
+    jintArray result = env->NewIntArray(size);
+    jint* buf = new jint[size];
+    for (int i = 0; i < size; ++i) {
+        buf[i] = productDetailsList[i]->hashCode;
```
Comment (qiuguohua) [inline native/cocos/platform/android/modules/google_play/billing/JniBillingHelper.cpp]: Finding objects in Java using hashCode

**Comment by qiuguohua** [comment]
@cocos-robot run test cases

File: cocos/core/platform/billing/billing.ts (hunk 1 above)
Comment (minggo) [reply cocos/core/platform/billing/billing.ts]: Yep, i think should remove it. But it is better to listen to the google guys' opinions too.

File: cocos/core/platform/billing/billing.ts (hunk 2 above)
Comment (minggo) [reply cocos/core/platform/billing/billing.ts]: Should only work on google play build target.

File: cocos/native-binding/index.ts [hunk 5]
```diff
@@ -1476,4 +1476,350 @@ export declare namespace native {
...
+    /**
+     * @en Represents the details of a one time or subscription product.
+     * @zh 代表一次性或订阅产品的详细信息。
+     */
+    export interface ProductDetails {
+        /**
+         * @en Represents the available purchase plans to buy a subscription product.
+         * @zh 代表一次性购买产品的报价详情。
+         */
+        readonly hashCode: number;
```
Comment (dumganhar) [inline cocos/native-binding/index.ts]: Is the comment for `hashCode` wrong?

File: cocos/native-binding/index.ts [hunk 6]
```diff
@@ -1476,4 +1476,350 @@ export declare namespace native {
...
+         * @zh ProductType产品的。
+         */
+        readonly productType: string;
+        /**
+         * @en The title of the product being sold.
+         * @zh 所售产品的标题。
+         */
+        readonly title: string;
+
+        readonly toStr: string;
```
Comment (dumganhar) [inline cocos/native-binding/index.ts]: Missing API comment for `toStr`

File: native/cocos/bindings/manual/jsb_billing.cpp [hunk 7]
```diff
@@ -0,0 +1,1461 @@
...
+    cls->defineFunction("showInAppMessages", _SE(js_cc_Billing_showInAppMessages));
+
+    obj->setProperty("billing", billingVal);
+
+    return true;
+}
+
+bool jsb_register_all_billing(se::Object* obj) { // NOLINT
+    se::Value nsVal;
+    if (!obj->getProperty("jsb", &nsVal, true)) {
```
Comment (dumganhar) [inline native/cocos/bindings/manual/jsb_billing.cpp:1471]: Should use another namespace, like `gp` or `googleplay`

**Pull request closed without merging**